cql3/expr/expression.hh

```diff
@@ -218,10 +217,15 @@ const column_value& get_subscripted_column(const subscript&);
 /// Only columns can be subscripted in CQL, so we can expect that the subscripted expression is a column_value.
 const column_value& get_subscripted_column(const expression&);
 
+/// Represents token(c1, c2) function on LHS of an operator relation.
+/// args contains arguments to the token function.
+struct token {
+    std::vector<expression> args;
 
-/// Represents token function on LHS of an operator relation.  No need to list column definitions
-/// here -- token takes exactly the partition key as its argument.
-struct token {};
+    explicit token(std::vector<expression>);
+    explicit token(const std::vector<const column_definition*>&);
+    explicit token(const std::vector<::shared_ptr<column_identifier_raw>>&);
```

Comment:
Do we need the last two constructors? The callers can convert their inputs to expressions.

If we remove them, we can remove the first constructor too.

Reply:
(there's no reason for token to presume anything about its parameters)

Reply:
we can have helpers make_token_expr_from(...) to ease the transition. Eventually the grammar will pass expressions there anyway.

Reply:
I feel like constructors are nicer than forcing the caller to create a vector of expressions.
With constructors the caller doesn't care what is the exact representation that token requires, they can just create it and move on. Converting is then neatly compartmentalized inside the constructor and caller can focus on their logic.
Additionally this eliminates repetition when other code will want to do the same thing.

I can add `make_token_expr_from(...)`, but is there any reason to have it instead of constructors? With constructors it's immediately obvious what they do and the function is easy to find for someone exploring the code. I feel like a free standing function would make it more complicated.

Reply:
My reasoning is to have a struct-like approach where it's plain data. But it's a minor point, we can keep it as is.

Reply:
Also I'll secretly hope that they become outdated when the grammar emits expressions for the arguments.